module/scheduler_components/simple_remaster_manager.cpp

```diff
@@ -0,0 +1,129 @@
+    return VerifyMasterResult::WAITING;
+  } else {
+    return result;
+  }
+}
+
+VerifyMasterResult SimpleRemasterManager::CheckCounters(TransactionHolder& txn_holder) {
+  auto& keys = txn_holder.KeysInPartition();
+  auto& txn_master_metadata = txn_holder.GetTransaction()->internal().master_metadata();
+  for (auto& key_pair : keys) {
+    auto key = key_pair.first;
+
+    // Get txn's counter
+    CHECK(txn_master_metadata.count(key))
+            << "Master metadata for key \"" << key << "\" is missing";
+    auto txn_counter = txn_master_metadata.at(key).counter();
+
+    // Get current counter from storage
+    auto storage_counter = 0; // default to 0 for a new key
+    Record record;
+    bool found = storage_->Read(key, record);
+    if (found) {        
+      storage_counter = record.metadata.counter;
+      CHECK(txn_master_metadata.at(key).master() == record.metadata.master)
+              << "Masters don't match for same key \"" << key << "\"";
+    }
+
+    if (txn_counter < storage_counter) {
+      return VerifyMasterResult::ABORT;
+    } else if (txn_counter > storage_counter) {
+      return VerifyMasterResult::WAITING;
+    }
+  }
+  return VerifyMasterResult::VALID;
+}
+
+RemasterOccurredResult
+SimpleRemasterManager::RemasterOccured(const Key remaster_key, const uint32_t remaster_counter) {
+  RemasterOccurredResult result;
+  // Try to unblock each txn at the head of a queue, if it contains the remastered key.
+  // Note that multiple queues could contain the same key with different counters
+  for (auto& queue_pair : blocked_queue_) {
+    auto txn_replica_id = queue_pair.second.front();
+    auto& txn_keys = all_txns_->at(txn_replica_id).KeysInPartition();
+    for (auto& key_pair : txn_keys) {
+      auto txn_key = key_pair.first;
+      if (txn_key == remaster_key) {
+        // TODO: check here if counters match, saves an iteration through all keys
+        TryToUnblock(queue_pair.first, result);
+        break;
+      }
+    }
+  }
+  return result;
+}
+
+void SimpleRemasterManager::TryToUnblock(
+    const uint32_t local_log_machine_id,
+    RemasterOccurredResult& result) {
+  if (blocked_queue_.count(local_log_machine_id) == 0) {
+    return;
+  }
+
+  auto txn_replica_id = blocked_queue_.at(local_log_machine_id).front();
+  auto& txn_holder = all_txns_->at(txn_replica_id);
+  auto& keys = txn_holder.KeysInPartition();
+
+  auto counter_result = CheckCounters(txn_holder);
+  if (counter_result == VerifyMasterResult::WAITING) {
+    return;
+  } else if (counter_result == VerifyMasterResult::VALID) {
+    result.unblocked.push_back(txn_replica_id);
+  } else if (counter_result == VerifyMasterResult::ABORT) {
+    result.should_abort.push_back(txn_replica_id);
+  }
+
+  // Head of queue has changed
+  blocked_queue_.at(local_log_machine_id).pop_front();
+  if (blocked_queue_.at(local_log_machine_id).empty()) {
+    // garbage collect
+    blocked_queue_.erase(local_log_machine_id);
```

Comment:
Number of replicas is usually small enough to keep these queues around so I think we should not erase them here to avoid recreating them.